KOGITO-1204 - BPMN editor to support "expression" in Data IO on process side

hasys opened the pull request:
Right now it is possible to create Output constant, save it and reopen the process.

Remains to fix:
- [x] Expressions should be saved without CDATA block and without encoding
- [x] Kogito marshallers
- [x] Find out better name for the "constant..."
- [x] add missing tests
- [x] sonar issues if any
- [x] fix popover
- [x] improve UI validation

hasys commented:
Jenkins retest please

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/InitializedVariable.java:
```diff
@@ -71,6 +71,24 @@ public static InitializedInputVariable inputOf(
     }
 
+    public static InitializedInputVariable createCustomInput(String parentId, VariableDeclaration varDecl, String expression) {
+        String isExpressionPattern = "^#\\{(.+)}$";
```
I think it is better to create the `Pattern` object as an `static` field in this class, This way you can just call `Pattern#match(s)` passing the test string, and the pattern is already pre-compiled only once.

hasys replied:
Good point, thanks

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/InitializedVariable.java:
```diff
@@ -71,6 +71,24 @@ public static InitializedInputVariable inputOf(
+    public static InitializedInputVariable createCustomInput(String parentId, VariableDeclaration varDecl, String expression) {
+        String isExpressionPattern = "^#\\{(.+)}$";
+        String decodedExpression = decode(expression);
+        if (decodedExpression.matches(isExpressionPattern)) {
```
The JIRA was expecting two options on the dropdown: `constant` and `expression`. So before merging, we should clarify if doing this way is still fine for eveyone.

hasys replied:
This is marshallers part we need to do it on server side anyway despite of how it will be handled on the UI part.

romartin replied:
got it, thx!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/InitializedVariable.java:
```diff
@@ -71,6 +71,24 @@ public static InitializedInputVariable inputOf(
+        try {
+            return URLDecoder.decode(text, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            throw new IllegalArgumentException(text, e);
```
By doing this you're converting the "checked" exception `UnsupportedEncodingException` into an "unchecked" one of type `IllegalArgumentException`. This results a bit "dangerous" because the signature for the `decode` method does not provide any exception information, but the decode can still fail, right? By doing this, actually it'll just thow an `IllegalArgumentException` and maybe it appears on the UI, but it's not really useful for the end user.
I was maybe expecting this decode method throwing the exception, and the caller for it catching the exception too, this way you can show a validation error to the user.
But again, not sure if makes completely sense, it depends on how to use it....

hasys replied:
This catch `UnsupportedEncodingException` since we have encoding as not changing literal it should work without issues.

romartin replied:
but I mean, if that `IllegalArgumentException` is being thrown at some point... what wil happen to the end user? does the exception will end up on the UI? does the field keep working properly? I'm just concerned about there this exception will end up....

romartin replied:
@hasys any comments here?

hasys replied:
Process won't be saved and in server log you will see:
<details>
  <summary>Click to expand</summary>
  <pre>ERROR 24-03 23:23:41,319 (BPMNStandaloneDiagramServiceImpl.java:transform:159) 	 Error whilst converting BPMN Diagram to XML.
java.lang.IllegalArgumentException: dsaf
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.decode(InitializedVariable.java:90)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.createCustomInput(InitializedVariable.java:78)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.inputOf(InitializedVariable.java:68)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.ParsedAssignmentsInfo.lambda$createInitializedInputVariables$0(ParsedAssignmentsInfo.java:191)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.ParsedAssignmentsInfo.createInitializedInputVariables(ParsedAssignmentsInfo.java:196)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.properties.ActivityPropertyWriter.setAssignmentsInfo(ActivityPropertyWriter.java:66)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.tasks.TaskConverter.userTask(TaskConverter.java:150)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$when$4(NodeMatch.java:85)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch$Case.match(NodeMatch.java:140)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$apply$5(NodeMatch.java:113)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.LinkedList$LLSpliterator.tryAdvance(LinkedList.java:1253)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.apply(NodeMatch.java:115)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.tasks.TaskConverter.toFlowElement(TaskConverter.java:64)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$when$4(NodeMatch.java:85)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch$Case.match(NodeMatch.java:140)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$apply$5(NodeMatch.java:113)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.LinkedList$LLSpliterator.tryAdvance(LinkedList.java:1253)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.apply(NodeMatch.java:115)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.FlowElementConverter.toFlowElement(FlowElementConverter.java:55)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.ProcessConverterDelegate.convertChildNodes(ProcessConverterDelegate.java:71)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverter.convertProcess(RootProcessConverter.java:52)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.DefinitionsConverter.toDefinitions(DefinitionsConverter.java:54)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.BaseDirectDiagramMarshaller.marshall(BaseDirectDiagramMarshaller.java:123)
	at deployment.business-central.war//org.kie.workbench.common.stunner.standalone.backend.services.BPMNStandaloneDiagramServiceImpl.transform(BPMNStandaloneDiagramServiceImpl.java:157)
	at deployment.business-central.war//org.kie.workbench.common.stunner.standalone.backend.services.BPMNStandaloneDiagramServiceImpl$Proxy$_$$_WeldClientProxy.transform(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at deployment.business-central.war//org.jboss.errai.bus.server.io.AbstractRPCMethodCallback.invokeMethodFromMessage(AbstractRPCMethodCallback.java:65)
	at deployment.business-central.war//org.jboss.errai.bus.server.io.ValueReplyRPCEndpointCallback.callback(ValueReplyRPCEndpointCallback.java:40)
	at deployment.business-central.war//org.jboss.errai.bus.server.io.RemoteServiceCallback.callback(RemoteServiceCallback.java:54)
	at deployment.business-central.war//org.jboss.errai.cdi.server.CDIExtensionPoints$2.callback(CDIExtensionPoints.java:448)
	at deployment.business-central.war//org.jboss.errai.bus.server.DeliveryPlan.deliver(DeliveryPlan.java:47)
	at deployment.business-central.war//org.jboss.errai.bus.server.ServerMessageBusImpl.sendGlobal(ServerMessageBusImpl.java:297)
	at deployment.business-central.war//org.jboss.errai.bus.server.SimpleDispatcher.dispatchGlobal(SimpleDispatcher.java:46)
	at deployment.business-central.war//org.jboss.errai.bus.server.service.ErraiServiceImpl.store(ErraiServiceImpl.java:96)
	at deployment.business-central.war//org.jboss.errai.bus.server.service.ErraiServiceImpl.store(ErraiServiceImpl.java:113)
	at deployment.business-central.war//org.jboss.errai.bus.server.servlet.DefaultBlockingServlet.doPost(DefaultBlockingServlet.java:144)
	at javax.servlet.api@1.0.0.Final//javax.servlet.http.HttpServlet.service(HttpServlet.java:706)
	at javax.servlet.api@1.0.0.Final//javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at deployment.business-central.war//org.kie.workbench.common.stunner.standalone.backend.CORSFilter.doFilter(CORSFilter.java:78)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.opentracing.contrib.opentracing-jaxrs2//io.opentracing.contrib.jaxrs2.server.SpanFinishingFilter.doFilter(SpanFinishingFilter.java:55)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.core@2.0.13.Final//io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.core@2.0.13.Final//io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.core@2.0.13.Final//io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.core@2.0.13.Final//io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.core@2.0.13.Final//io.undertow.server.Connectors.executeRootHandler(Connectors.java:360)
	at io.undertow.core@2.0.13.Final//io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads@2.3.2.Final//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1985)
	at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1487)
	at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1378)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.UnsupportedEncodingException: UTF-
	at java.base/java.net.URLDecoder.decode(URLDecoder.java:144)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.decode(InitializedVariable.java:88)
	... 121 more
</pre>
</details>

Also it will be caught by failing test:
<details>
  <summary>InitializedVariableTest#urlDecodeConstants</summary>
  <pre>java.lang.IllegalArgumentException: %3C%3C%3C%23%21%21%21%23%3E%3E%3E

	at org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.decode(InitializedVariable.java:90)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.createCustomInput(InitializedVariable.java:78)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.properties.InitializedVariableTest.urlDecodeConstants(InitializedVariableTest.java:53)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: java.io.UnsupportedEncodingException: UTF-
	at java.lang.StringCoding.decode(StringCoding.java:190)
	at java.lang.String.<init>(String.java:426)
	at java.net.URLDecoder.decode(URLDecoder.java:190)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.decode(InitializedVariable.java:88)
	... 25 more</pre>
</details>

I think it safe enough, since to get this error you need to modify and recompile the code (no user will do it) and also change literal to incorrect value and ignore the failing test.

hasys replied:
We can ignore this exception and return empty value, for an instance and in this case in worse case user will lose only value for this property, but it will be very hard to debug.

hasys replied:
Anyway user can do nothing to fix incorrect literal, because to fix it you need to modify the code and rebuild business central. So I am not sure how useful fancy message will be for the end user if someone will change for any reason this literal to incorrect one.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/tostunner/properties/OutputAssignmentReader.java:
```diff
@@ -50,4 +79,18 @@ public AssociationDeclaration getAssociationDeclaration() {
+                    try {
+                        return URLEncoder.encode(b, "UTF-8");
+                    } catch (UnsupportedEncodingException e) {
+                        throw new IllegalArgumentException(body, e);
```
maybe here happens the same issue as on the `decode` method above - around un/checked exceptions..

hasys replied:
This catch `UnsupportedEncodingException` since we have encoding as not changing literal it should work without issues.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/converters/fromstunner/properties/InitializedVariableTest.java:
```diff
@@ -66,6 +69,19 @@ public void setup() {
 
+    @Test
+    public void testInputCustomVariable() {
+        VariableDeclaration declaration = mock(VariableDeclaration.class);
```
Is there really a need to mock the `VariableDeclaration` object? In general, for domain objects (with no logic around) there is no need to mock the instances, it just makes the test more complex...
If possible, please try to use the `new` operator, create the "real" instance, and if necessary, just `spy` on it.

hasys replied:
Make sense, thank you!

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins retest please

romartin commented:
hey @hasys - can me and @domhanak proceed with that PR reivew? if so, why not moving from draft to a normal PR?

hasys commented:
Hi @romartin, @lazarotti, @domhanak, @MarianMacik,

yes, it should work, but when I test it with VS Code extension it is only Input expressions works, Output expressions doesn't. XML looks the same to me as it is in [engine tests](https://github.com/kiegroup/kogito-runtimes/pull/314/files?file-filters%5B%5D=.bpmn2#diff-3284b19d7b5e8e81d79bf6ceb4eb011b) (see below).

I used this quickstart: https://github.com/kiegroup/kogito-quickstarts/tree/master/kogito-usertasks-quarkus it has latest Kogito version so it should work. (to test it don't use `admin` user as described, but `manager` user. See https://github.com/kiegroup/kogito-quickstarts/pull/15.

### XML tested on engine side:
https://github.com/kiegroup/kogito-runtimes/pull/314/files?file-filters%5B%5D=.bpmn2#diff-3284b19d7b5e8e81d79bf6ceb4eb011bR71
```xml
<bpmn2:dataOutputAssociation id="DataOutputAssociation_1">
    <bpmn2:sourceRef>DataOutput_1</bpmn2:sourceRef>
    <bpmn2:assignment id="Assignment_1">
        <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="FormalExpression_2">DataOutput_1</bpmn2:from>
        <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="FormalExpression_4">#{person.age}</bpmn2:to>
    </bpmn2:assignment>
</bpmn2:dataOutputAssociation>
```

### XML generated by Stunner
When I changed `ActorId`->`approver` output variable to `ActorId`->`#{approver}` it generates the same XML (IDs are different because Maciej generated example in Eclipse and names of variables are different because it is a different process)
```xml
<bpmn2:dataOutputAssociation id="_toMzAUuCEDiDBqzT9zOfzw">
    <bpmn2:sourceRef>_8B62D3CA-5D03-4B2B-832B-126469288BB4_ActorIdOutputX</bpmn2:sourceRef>
    <bpmn2:assignment id="_toMzAkuCEDiDBqzT9zOfzw">
        <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="_toNaEEuCEDiDBqzT9zOfzw">_8B62D3CA-5D03-4B2B-832B-126469288BB4_ActorIdOutputX</bpmn2:from>
        <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="_toNaEUuCEDiDBqzT9zOfzw">#{approver}</bpmn2:to>
    </bpmn2:assignment>
</bpmn2:dataOutputAssociation>
```

So, not sure is it an issue in engine or I did something wrong. Will continue my investigations for now. Please let me know if you have any idea.

P.S. I fixed failed test and rebased to master.

hasys commented:
Jenkins execute full downstream build

lazarotti commented:
@hasys just to be here what we talked, this feature is for complex types and will not work for literals. Literals are direct set using the widget (Target in dataoutput) and there is no need to use expressions.

hasys commented:
Thank you @lazarotti, it did a trick. I used for tests complex object in Input and String in output, that's why Outputs didn't work.

I also added fix for [KOGITO-1209](https://issues.redhat.com/browse/KOGITO-1209) since it is the same lines of code as in this PR.

The only issue is CDATA block is not saved in Kogito marshallers, but I did the same call to CData generation as everywhere else so if it will work in other places it will work there.

I tested CDATA block with engine as well (by manually adding CDATA blocks and executing examples).

hasys commented:
Hi @domhanak, @romartin,

to test this PR at runtime you can use https://github.com/kiegroup/kogito-quickstarts/tree/master/kogito-usertasks-quarkus see below for step by step guide. I think this PR is ready for review, but if you think it will be better to review it with fixed CDATA issue, it is fine, but you already can start Code review. Thank you!

#### In the process modify *First Line Approval*'s data assignments like this:
![image](https://user-images.githubusercontent.com/1477262/77026260-86267e00-6993-11ea-8908-032af928520f.png)
Input `traveller<-#{traveller}`
Output `ActorId->#{traveller.firstName}`

#### Run example
use `mvn clean package quarkus:dev -DskipTests` to run the example (test is failing because of our modification, but we don't need it we want to test End-to-End runtime)


#### Check existing requests by executing
```
curl -X GET "http://localhost:8080/approvals" -H  "accept: application/json"
```
Empty value should be found

#### Than add new approval request by
```
curl -X POST "http://localhost:8080/approvals" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"traveller\":{\"firstName\":\"John\",\"lastName\":\"Doe\",\"email\":\"jon.doe@example.com\",\"nationality\":\"American\",\"address\":{\"street\":\"main street\",\"city\":\"Boston\",\"zipCode\":\"10005\",\"country\":\"US\"}}}"
```

#### Check that request was added
```
curl -X GET "http://localhost:8080/approvals" -H  "accept: application/json"

[{"id":"1763e6a9-b423-4beb-9ad6-d6edc6960e29","approver":null,"firstLineApproval":null,"secondLineApproval":null,"traveller":{"firstName":"John","lastName":"Doe","email":"jon.doe@example.com","nationality":"American","address":{"street":"main street","city":"Boston","zipCode":"10005","country":"US"}}}]
```

#### Copy ID to the next request
```
curl -X GET "http://localhost:8080/approvals/{PUT_YOUR_ID_HERE}/tasks?user=manager" -H  "accept: application/json"

{"a34699b1-fa96-489e-84df-7e7075091fc2":"firstLineApproval"}
```

#### Copy second ID for the next request:
```
curl -X POST "http://localhost:8080/approvals/{PUT_YOUR_FIRST_ID_HERE}/firstLineApproval/{PUT_YOUR_SECOND_ID_HERE}?phase=complete&user=manager" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"approved\":true}"

{
  "id": "1763e6a9-b423-4beb-9ad6-d6edc6960e29",
  "approver": null,
  "firstLineApproval": true,
  "secondLineApproval": null,
  "traveller": {
    "firstName": "manager",
    "lastName": "Doe",
    "email": "jon.doe@example.com",
    "nationality": "American",
    "address": {
      "street": "main street",
      "city": "Boston",
      "zipCode": "10005",
      "country": "US"
    }
  }
}
```

#### Result
Now you can see two things: Traveller is correctly moved to the *First Line Approval* since it is updated. *firstName* property of the Traveller is updated according to our changes in the process (**manager** instead of **John**).

hasys commented:
I rebased to the master and fixed all conflicts.

Jenkins execute full downstream build

hasys commented:
SonarCloud failed due to duplications on marshallers side.

romartin commented:
oh great job @hasys !  Thnks for sharing the [above information](https://github.com/kiegroup/kie-wb-common/pull/3170#issuecomment-600958885), it helps. Also thanks for fixing the related issues around this too :+1:

hasys commented:
Hi @domhanak, @romartin,

CDATA generates correctly in web app, in VS Code it is not added due to prettifier used in Kogito-tooling, it is not related to Stunner. So CDATA can be tested in Web-app for now. Thank you!

domhanak commented:
I tried following scenario on BC, but I am not sure if it is suported:
1. Create a process with (start -> HT -> end)
2. Create a process Variable of type Person(with name only)
3. In human task create data I/O assignments like this:
![Screenshot from 2020-03-24 14-37-12](https://user-images.githubusercontent.com/5937251/77431555-25f16b00-6ddd-11ea-8d86-a7f97fad4e08.png)
4. Generate Forms and Build & Deploy
5. Start process instance, input name.
6. Navigate to Task Inbox and process you task. The form looks like this:
![Screenshot from 2020-03-24 14-46-24](https://user-images.githubusercontent.com/5937251/77432397-4b32a900-6dde-11ea-9e47-1c463d19eda9.png)
As you can see it correctly show the name of the person using expression, but the Object itself doesn't show the name field as it should.

I discussed with @hasys but I need wider audience to asses this scenario.

hasys commented:
Hi @domhanak,

great catch, I reported it to jBPM engine: [JBPM-9072](https://issues.redhat.com/browse/JBPM-9072).

For now we can ensure that XML generated correctly. There are my findings:

#### XML for input variables generated in Eclipse plugin:
```xml
      <bpmn2:dataInputAssociation id="DataInputAssociation_9">
        <bpmn2:targetRef>DataInput_3</bpmn2:targetRef>
        <bpmn2:assignment id="Assignment_8">
          <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="FormalExpression_18">#{person.name}</bpmn2:from>
          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="FormalExpression_16">DataInput_3</bpmn2:to>
        </bpmn2:assignment>
      </bpmn2:dataInputAssociation>
```
#### XML for input variables generated in Business Central:
```xml
      <bpmn2:dataInputAssociation id="_-3pVxG3dEeqg_N33Q3MbyQ">
        <bpmn2:targetRef>_2_personNameInputX</bpmn2:targetRef>
        <bpmn2:assignment id="_-3pVxW3dEeqg_N33Q3MbyQ">
          <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="_-3pVxm3dEeqg_N33Q3MbyQ"><![CDATA[#{person.name}]]></bpmn2:from>
          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="_-3pVx23dEeqg_N33Q3MbyQ">_2_personNameInputX</bpmn2:to>
        </bpmn2:assignment>
      </bpmn2:dataInputAssociation>
```

#### XML for output variables generated in Eclipse plugin:
```xml
      <bpmn2:dataOutputAssociation id="DataOutputAssociation_2">
        <bpmn2:sourceRef>DataOutput_1</bpmn2:sourceRef>
        <bpmn2:assignment id="Assignment_9">
          <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="FormalExpression_19">DataOutput_1</bpmn2:from>
          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="FormalExpression_21">#{person.name}</bpmn2:to>
        </bpmn2:assignment>
      </bpmn2:dataOutputAssociation>
```

#### XML for output variables generated in Business Central:
```xml
      <bpmn2:dataOutputAssociation id="_-3pVzG3dEeqg_N33Q3MbyQ">
        <bpmn2:sourceRef>_2_personAgeOutputX</bpmn2:sourceRef>
        <bpmn2:assignment id="_-3pVzW3dEeqg_N33Q3MbyQ">
          <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="_-3pVzm3dEeqg_N33Q3MbyQ">_2_personAgeOutputX</bpmn2:from>
          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="_-3pVz23dEeqg_N33Q3MbyQ"><![CDATA[#{person.age}]]></bpmn2:to>
        </bpmn2:assignment>
      </bpmn2:dataOutputAssociation>
```

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/InitializedVariable.java:
```diff
@@ -38,6 +40,7 @@
     private final String identifier;
     private final String type;
     private ItemDefinition itemDefinition;
+    private static final Pattern EXPRESSION = Pattern.compile(Patterns.EXPRESSION);
```
why not moving this `static final Pattern` instance to the new `Patterns` class you just created? So htis way, the expression (`String`) will be compiled just once, instead of creating a new `Pattern` instance in every place you need it, and so also having to compile the string expression too.

hasys replied:
It is because I can't reuse Pattern in GWT and classic java pattern, both have different implementations. It looks like the String is the only thing I can keep in shared code.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/util/Patterns.java:
```diff
@@ -0,0 +1,28 @@
+ */
+public class Patterns {
+
+    public static final String EXPRESSION = "^#\\{(.+)}$";
```
I would just use a `Pattern` instance here as well, so the expression will be only compiled once. See above comments too about this.

hasys replied:
It is because I can't reuse Pattern in GWT and classic java pattern, both have different implementations. It looks like the String is the only thing I can keep in shared code.

romartin replied:
ahh good point, it's a "shared" class sorry, I didn't noticed before. You're right, all good @hasys thanks!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/test/java/org/kie/workbench/common/stunner/bpmn/backend/converters/tostunner/properties/OutputAssignmentReaderTest.java:
```diff
@@ -53,7 +58,35 @@ public void testFromAssociationWithPropertyId() {
+
+    @Test
+    public void testFromAssociationWithExpression() {
+        DataOutput output = spy(DataOutput.class);
```
instead of mocking, it's better to use "real" model objects on tests, if possible.

hasys replied:
Yes, I agree, but it has protected constructor and it is not possible to instantiate DataOutputImpl from our tests.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/assignmentsEditor/ActivityDataIOEditorWidget.html:
```diff
@@ -23,29 +23,9 @@
     <meta name="description" content="">
     <meta name="author" content="">
 
-    <!--link rel="stylesheet/less" href="less/bootstrap.less" type="text/css" /-->
```
good catch. In fact, for all these widgets that are using `html` templates I'm pretty sure the whole `head` can be removed, also no need for `body` tag... it's because those html represents "fragments" of html files, which are alrady appended to complete documents, so you can feel free also to remove the `head`/`body` tags as well.

hasys replied:
Sounds good, will remove thank you

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/assignmentsEditor/AssignmentListItemWidgetViewImpl.java:
```diff
@@ -274,13 +275,24 @@ public void setProcessVar(final String processVar) {
+                                                    NotificationEvent.NotificationType.ERROR));
+            processVarComboBox.textBoxValueChanged("");
+
+            return;
```
minor comment  - maybe results more clear to use `if / else` instead of adding a `return` keyword here...

hasys replied:
Good catch, thanks

hasys commented:
Hi @domhanak, changes after code review are very small and isolated, you can see it in last commit. It is shared UI code so it works the same for both Business Central and Kogito.

I tested it on Business Central standalone showcase, all works as before. So you can continue your testing in Kogito. Thank you!

hasys commented:
Jenkins execute full downstream build

domhanak commented:
Output association in VSCode after saveing process:
```
 <bpmn2:dataOutputAssociation id="_aUvmtlCjEDibiNhQmZqG_g">
        <bpmn2:sourceRef>_86AA549D-2ECA-4B58-A805-1E7749F8B6D1_approvedOutputX</bpmn2:sourceRef>
        <bpmn2:assignment id="_aUwNwFCjEDibiNhQmZqG_g">
          <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="_aUwNwVCjEDibiNhQmZqG_g">_86AA549D-2ECA-4B58-A805-1E7749F8B6D1_approvedOutputX</bpmn2:from>
          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="_aUwNwlCjEDibiNhQmZqG_g">#{approval}</bpmn2:to>
        </bpmn2:assignment>
      </bpmn2:dataOutputAssociation>
```

Input association:

```
<bpmn2:dataInputAssociation id="_aUu_oVCjEDibiNhQmZqG_g">
        <bpmn2:targetRef>_86AA549D-2ECA-4B58-A805-1E7749F8B6D1_personOkInputX</bpmn2:targetRef>
        <bpmn2:assignment id="_aUu_olCjEDibiNhQmZqG_g">
          <bpmn2:from xsi:type="bpmn2:tFormalExpression" id="_aUvmsFCjEDibiNhQmZqG_g">#{person}</bpmn2:from>
          <bpmn2:to xsi:type="bpmn2:tFormalExpression" id="_aUvmsVCjEDibiNhQmZqG_g">_86AA549D-2ECA-4B58-A805-1E7749F8B6D1_personOkInputX</bpmn2:to>
        </bpmn2:assignment>
      </bpmn2:dataInputAssociation>
```

This looks good to me,  I just need to run the example now.

hasys commented:
Hi @romartin, this PR is ready for merge. Thank you!
